Support sp_rename for OBJECT type

raydhkim opened the pull request:
Previously, Babelfish did not support sp_rename for Babelfish objects, which prevented the users from renaming objects and table columns. It is important to support sp_rename to allow users change and modify their own-created objects in the database. Now, sp_rename for OBJECT type is supported, allowing users to rename their own-created TABLE, VIEW, PROCEDURE, FUNCTION.

Task: BABEL-3888

Signed-off-by: Ray Kim <raydhkim@amazon.com>

### Test Scenarios Covered ###
* **Use case based -**


* **Boundary conditions -**


* **Arbitrary inputs -**


* **Negative test cases -**


* **Minor version upgrade tests -**


* **Major version upgrade tests -**


* **Performance tests -**


* **Tooling impact -**


* **Client tests -**



### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

raydhkim committed:
Support sp_rename for OBJECT type

Previously, Babelfish did not support sp_rename for Babelfish objects,
which prevented the users from renaming objects and table columns. It is
important to support sp_rename to allow users change and modify their
own-created objects in the database. Now, sp_rename for OBJECT type is
supported, allowing users to rename their own-created TABLE, VIEW,
PROCEDURE, FUNCTION.

Task: BABEL-3888

Signed-off-by: Ray Kim <raydhkim@amazon.com>

macwoody commented on contrib/babelfishpg_tsql/sql/babelfishpg_tsql.sql:
```diff
@@ -2903,3 +2903,62 @@ END;
+	IN "@newname" sys.SYSNAME,
+	IN "@schemaname" sys.nvarchar(776),
+	IN "@objtype" char(2) DEFAULT NULL
+) AS 'babelfishpg_tsql', 'sp_rename_internal' LANGUAGE C;
```
also need to grant to this internal one

macwoody commented on contrib/babelfishpg_tsql/sql/babelfishpg_tsql.sql:
```diff
@@ -2903,3 +2903,62 @@ END;
+LANGUAGE 'pltsql'
+AS $$
+BEGIN
+	IF @objtype IS NOT NULL AND (@objtype != 'OBJECT')
```
plz check object is null case

macwoody commented on contrib/babelfishpg_tsql/sql/babelfishpg_tsql.sql:
```diff
@@ -2903,3 +2903,62 @@ END;
+	DECLARE @subname sys.nvarchar(776) = '';
+	DECLARE @schemaname sys.nvarchar(776) = '';
+	SELECT @name_count = COUNT(*) FROM STRING_SPLIT(@objname, '.');
+	IF @name_count > 1
```
what if you have 3 part, like 
master.dbo.t1 ?

macwoody commented on contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--2.3.0--2.4.0.sql:
```diff
@@ -1304,4 +1362,4 @@ LANGUAGE C STABLE;
 
 -- Reset search_path to not affect any subsequent scripts
-SELECT set_config('search_path', trim(leading 'sys, ' from current_setting('search_path')), false);
+SELECT set_config('search_path', trim(leading 'sys, ' from current_setting('search_path')), false);
```
what is this change ?

macwoody commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -3117,18 +3117,56 @@ static void bbf_ProcessUtility(PlannedStmt *pstmt,
+				dbo_id = get_role_oid(dbo_name, false);
+
+				/* Set current user to dbo for alter permissions */
+				prev_current_user = GetUserId();
```
why do we force switch to dbo ? 
this will allow any user to rename object

macwoody commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -3117,18 +3117,56 @@ static void bbf_ProcessUtility(PlannedStmt *pstmt,
+				PG_END_TRY();
+				/* Clean up. Restore previous state. */
+				SetCurrentRoleId(prev_current_user, false);
 				return;
```
pay attention to this line , you did break

raydhkim committed:
Support sp_rename for OBJECT

Previously, Babelfish did not support sp_rename for Babelfish objects,
which prevented the users from renaming objects and table columns. It is
important to support sp_rename to allow users change and modify their
own-created objects in the database. Now, sp_rename for OBJECT type is
supported, allowing users to rename their own-created TABLE, VIEW,
PROCEDURE, FUNCTION.

Task: BABEL-3888

Signed-off-by: Ray Kim <raydhkim@amazon.com>

raydhkim committed:
Revise test case for sp_rename

raydhkim committed:
Merge remote-tracking branch 'upstream/BABEL_3_X_DEV' into jira-babel-3888-3xdev

raydhkim committed:
Merge remote-tracking branch 'upstream/BABEL_3_X_DEV' into jira-babel-3888-3xdev

raydhkim committed:
Make vu-test and dep-vu-test more distinguishable

Signed-off-by: Ray Kim <raydhkim@amazon.com>

raydhkim committed:
Merge remote-tracking branch 'upstream/BABEL_3_X_DEV' into jira-babel-3888-3xdev

raydhkim committed:
Revise upgrade script

Signed-off-by: Ray Kim <raydhkim@amazon.com>
